compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/transformers/body/resolve/FirControlFlowStatementsResolveTransformer.kt

```diff
@@ -216,9 +216,12 @@ class FirControlFlowStatementsResolveTransformer(transformer: FirBodyResolveTran
     ): CompositeTransformResult<FirStatement> {
         if (elvisExpression.calleeReference is FirResolvedNamedReference) return elvisExpression.compose()
         elvisExpression.transformAnnotations(transformer, data)
-        elvisExpression.transformLhs(transformer, ResolutionMode.ContextDependent)
+        val expectedArgumentType =
+            if (data is ResolutionMode.WithExpectedType && data.expectedType !is FirImplicitTypeRef) data
```

Comment:
Not sure what makes things look questionable, but it wouldn't be a coincidence that top-down propagation of types (from explicit info of enclosing one to inner item) appears again at https://github.com/JetBrains/kotlin/pull/3968/commits/3815dca0a0298b3698a0ae1b911d8030de4266ac